lib/SIL/SILFunctionType.cpp

```diff
@@ -340,124 +450,232 @@ CanSILFunctionType SILFunctionType::getAutoDiffDerivativeFunctionType(
+    auto paramTan =
+        param.getInterfaceType()->getAutoDiffTangentSpace(lookupConformance);
+    assert(paramTan && "Parameter type does not have a tangent space?");
+    auto paramTanType = paramTan->getCanonicalType();
+    auto resultTanConvention = getTangentResultConventionForOriginalParameter(
+        paramTanType, param.getConvention());
+    if (!paramTanType->hasArchetype() && !paramTanType->hasTypeParameter()) {
+      pullbackResults.push_back({paramTanType, resultTanConvention});
     } else {
-      auto &origRes = originalResults[resultIndex];
-      auto resultTan = origRes.getInterfaceType()->getAutoDiffTangentSpace(
-          lookupConformance);
-      assert(resultTan && "Result type does not have a tangent space?");
-      pullbackParams.push_back(getTangentParameterInfoForOriginalResult(
-          resultTan->getCanonicalType(), origRes.getConvention()));
-    }
-    SmallVector<SILResultInfo, 8> pullbackResults;
-    for (auto &param : diffParams) {
-      if (param.isIndirectInOut())
-        continue;
-      auto paramTan =
-          param.getInterfaceType()->getAutoDiffTangentSpace(lookupConformance);
-      assert(paramTan && "Parameter type does not have a tangent space?");
-      pullbackResults.push_back(getTangentResultInfoForOriginalParameter(
-          paramTan->getCanonicalType(), param.getConvention()));
-    }
-    closureType = SILFunctionType::get(
-        /*genericSignature*/ nullptr, ExtInfo(), SILCoroutineKind::None,
-        ParameterConvention::Direct_Guaranteed, pullbackParams, {},
-        pullbackResults, {}, getCombinedSubstitutions(), SubstitutionMap(),
-        ctx);
-    break;
+      auto gpIndex = substGenericParams.size();
+      auto gpType = CanGenericTypeParamType::get(0, gpIndex, ctx);
+      substGenericParams.push_back(gpType);
+      substReplacements.push_back(paramTanType);
+      pullbackResults.push_back({gpType, resultTanConvention});
+    }
   }
+  SubstitutionMap substitutions;
+  if (!substGenericParams.empty()) {
+    auto genericSig =
+        GenericSignature::get(substGenericParams, substRequirements)
+            .getCanonicalSignature();
+    substitutions =
+        SubstitutionMap::get(genericSig, llvm::makeArrayRef(substReplacements),
+                             llvm::makeArrayRef(substConformances));
   }
+  return SILFunctionType::get(
+      GenericSignature(), SILFunctionType::ExtInfo(), SILCoroutineKind::None,
+      ParameterConvention::Direct_Guaranteed, pullbackParams, {},
+      pullbackResults, None, substitutions,
+      /*invocationSubstitutions*/ SubstitutionMap(), ctx);
+}
+
+/// Constrains the `original` function type according to differentiability
+/// requirements:
+/// - All differentiability parameters are constrained to conform to
+///   `Differentiable`.
+/// - The invocation generic signature is replaced by the
+///   `constrainedInvocationGenSig` argument.
+static SILFunctionType *getConstrainedAutoDiffOriginalFunctionType(
```

Comment:
Note: maybe we can inline this helper function in `getAutoDiffDerivativeFunctionType` to avoid the overhead of constructing a new `SILFunctionType`.

The helper's primary purpose is to call `autodiff::getConstrainedDerivativeGenericSignature`. It also maps parameters and results using the constrained derivative generic signature, but callers could do that too.